### PR TITLE
[issue-2283]Extending coverage for 2283 issue [internal review]

### PR DESCRIPTION
**Pull request by VjKlepikov**
ATF Test Scripts to check [FORDTCN-11679](https://adc.luxoft.com/jira/browse/FORDTCN-11679)

This PR is **[ready]** for review.

### Summary
Extending coverage for 2238 issue

### ATF version
develop

### Changelog
Added new scripts
Updated test set for vehicle_data with new scripts

##### Enhancements
* [Enhancement info]

### CLA
- [X] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: test_scripts/API/VehicleData/common.lua [hunk 1]
```diff
@@ -325,6 +325,24 @@ function m.getVDParams(pIsSubscribable)
...
+    if value ~= "" then
+      if param ~= pParam then
+        out = param
+        break
+      end
+    end
+  end
+  return out
+end
+
```
Comment (dboltovskyi) [inline test_scripts/API/VehicleData/common.lua:345]: I would change this function to the following:
```
  local params = m.getVDParams(true)
  local sortedParams = {}
  for k in m.spairs(params) do
    table.insert(sortedParams, k)
  end
  for i, p in pairs(sortedParams) do
    if p == pParam then
      if i == #sortedParams then
        return sortedParams[1]
      else
        return sortedParams[i+1]
      end
    end
  end
```
It allows to remove randomization of VD parameters combinations by returning next parameter after defined one.
If defined parameter is the latest one, function returns 1st.

File: test_scripts/API/VehicleData/common.lua [hunk 2]
```diff
@@ -455,17 +473,83 @@ function m.unexpectedDisconnect(pParam)
   utils.wait(1000)
 end
 
+--[[ @unexpectedDisconnect: Unexpected disconnect sequence for two apps for different <vd_param> parameters
+--! @parameters:
+--! pParam1: name of the VD parameter for 1st app
+--! pParam2: name of the VD parameter for 2st app
+--! @return: none
+--]]
+function m.unexpectedDisconnectTwoApps(pParam1, pParam2)
```
Comment (dboltovskyi) [inline test_scripts/API/VehicleData/common.lua:482]: Name of the function doesn't correspond to name in description.

File: test_scripts/API/VehicleData/common.lua [hunk 3]
```diff
@@ -455,17 +473,83 @@ function m.unexpectedDisconnect(pParam)
...
+  actions.mobile.disconnect()
+  utils.wait(1000)
+end
+
+--[[ @unregisterApp: Unregister App sequence
+--! @parameters:
+--! pAppId: application number (1, 2, etc.)
+--! @return: none
+--]]
+function m.unregisterAppWithUnsubscription(pParam, pAppId, isRequestOnHMIExpected)
```
Comment (dboltovskyi) [inline test_scripts/API/VehicleData/common.lua:504]: Name of the function doesn't correspond to name in description.

File: test_scripts/API/VehicleData/common.lua [hunk 4]
```diff
@@ -455,17 +473,83 @@ function m.unexpectedDisconnect(pParam)
...
+    end)
+  else
+    m.getHMIConnection():ExpectRequest("VehicleInfo.UnsubscribeVehicleData"):Times(0)
+  end
+  local mobileSession = m.getMobileSession(pAppId)
+  local cid = mobileSession:SendRPC("UnregisterAppInterface",{})
+  actions.getHMIConnection():ExpectNotification("BasicCommunication.OnAppUnregistered",
+    { unexpectedDisconnect = false })
+  mobileSession:ExpectResponse(cid, { success = true, resultCode = "SUCCESS"})
+end
```
Comment (dboltovskyi) [inline test_scripts/API/VehicleData/common.lua:526]: Can we use existing `processSubscriptionRPC()` inside of this function in order not to have code duplicates?

File: test_scripts/API/VehicleData/common.lua [hunk 5]
```diff
@@ -455,17 +473,83 @@ function m.unexpectedDisconnect(pParam)
...
+      return true
+    else
+      return false, "VehicleInfo.UnsubscribeVehicleData request contains unexpected parameter" ..
+        utils.tableToString(data.params)
+    end
+  end)
+  :Times(2)
+  actions.mobile.disconnect()
+  utils.wait(1000)
+end
```
Comment (dboltovskyi) [inline test_scripts/API/VehicleData/common.lua:497]: Can we update existing `unexpectedDisconnect()` function as we did for `ignitionOff()`?

File: test_sets/API/vehicle_data.txt [hunk 6]
```diff
@@ -36,3 +38,6 @@
 ./test_scripts/API/VehicleData/UnsubscribeVehicleData/006_UnsubscribeVD_2nd_request_IGNORED.lua
 ./test_scripts/API/VehicleData/UnsubscribeVehicleData/007_UnsubscribeVD_for_2_apps.lua
 ./test_scripts/API/VehicleData/UnsubscribeVehicleData/008_UnsubscribeVD_HMI_responds_with_not_success_codes.lua
+./test_scripts/API/VehicleData/UnsubscribeVehicleData/009_UnsubscribeVD_for_2_apps_different_data.lua
+./test_scripts/API/VehicleData/UnsubscribeVehicleData/010_UnsubscribeVD_for_2_apps_unregister_data.lua
+./test_scripts/API/VehicleData/UnsubscribeVehicleData/011_UnsubscribeVD_for_2_apps_unregister_different_data.lua
```
Comment (dboltovskyi) [inline test_sets/API/vehicle_data.txt:43]: These 3 scripts for `UnsubscribeVehicleData` have sequences that are not related to the original issue. Do we really need them?

File: test_sets/API/vehicle_data.txt [hunk 7]
```diff
@@ -36,3 +38,6 @@
 ./test_scripts/API/VehicleData/UnsubscribeVehicleData/006_UnsubscribeVD_2nd_request_IGNORED.lua
 ./test_scripts/API/VehicleData/UnsubscribeVehicleData/007_UnsubscribeVD_for_2_apps.lua
 ./test_scripts/API/VehicleData/UnsubscribeVehicleData/008_UnsubscribeVD_HMI_responds_with_not_success_codes.lua
+./test_scripts/API/VehicleData/UnsubscribeVehicleData/009_UnsubscribeVD_for_2_apps_different_data.lua
+./test_scripts/API/VehicleData/UnsubscribeVehicleData/010_UnsubscribeVD_for_2_apps_unregister_data.lua
```
Comment (dboltovskyi) [inline test_sets/API/vehicle_data.txt:42]: Looks like the same sequence is covered in existing script: `007_UnsubscribeVD_for_2_apps.lua`

**Review by dboltovskyi — CHANGES_REQUESTED** [review]

File: test_scripts/API/VehicleData/common.lua (hunk 4 above)
Comment (VjKlepikov) [reply test_scripts/API/VehicleData/common.lua:518]: I suppose not due to redundant UnsubscribeVehicleData RPC in existing processSubscriptionRPC () 
local cid = m.getMobileSession(pAppId):SendRPC(pRPC, { [pParam] = true })

**Comment by VjKlepikov** [comment]
@dboltovskyi Thank you for the review, all comments have been processed 
Please note that added logic to skip test in case restricte VD params #VD_PARAMS=1  [a715e03](https://github.com/LuxoftSDL/sdl_atf_test_scripts/pull/114/commits/a715e03dd61e7af24fa345dc40d46f5b9e768380)

File: test_scripts/API/VehicleData/common.lua [hunk 8]
```diff
@@ -442,30 +465,83 @@ end
...
---! pParam: name of the VD parameter
+--! pParam1: name of the VD parameter
+--! pParam2: name of the VD parameter
 --! @return: none
 --]]
-function m.ignitionOff(pParam)
+function m.ignitionOff(pParam1, pParam2)
+  local expTimes = 0
+  if pParam1 and pParam2 then expTimes = 2
+  else if pParam1 or pParam2 then expTimes = 1 end end
```
Comment (dboltovskyi) [inline test_scripts/API/VehicleData/common.lua:530]: I would split this into a few lines since `end end` looks odd

File: test_scripts/API/VehicleData/common.lua [hunk 9]
```diff
@@ -442,30 +465,83 @@ end
...
---! pParam: name of the VD parameter
+--! pParam1: name of the VD parameter
+--! pParam2: name of the VD parameter
 --! @return: none
 --]]
-function m.unexpectedDisconnect(pParam)
+function m.unexpectedDisconnect(pParam1, pParam2)
+  local expTimes = 0
+  if pParam1 and pParam2 then expTimes = 2
+  else if pParam1 or pParam2 then expTimes = 1 end end
```
Comment (dboltovskyi) [inline test_scripts/API/VehicleData/common.lua:475]: I would split this into a few lines since `end end` looks odd

File: test_scripts/API/VehicleData/UnsubscribeVehicleData/011_UnsubscribeVD_for_2_apps_Expected_Disconnect_different_data.lua [hunk 10]
```diff
@@ -0,0 +1,64 @@
+---------------------------------------------------------------------------------------------------
+---------------------------------------------------------------------------------------------------
```
Comment (dboltovskyi) [inline test_scripts/API/VehicleData/UnsubscribeVehicleData/011_UnsubscribeVD_for_2_apps_Expected_Disconnect_different_data.lua:2]: Extra line with `---`

File: test_scripts/API/VehicleData/UnsubscribeVehicleData/011_UnsubscribeVD_for_2_apps_Expected_Disconnect_different_data.lua [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+-- 4) Apps are subscribed to <vd_param> parameter data
+--
+-- In case:
+-- 1) App_1 sends UnregisterAppInterface request to SDL
+-- SDL does:
+-- a) send UnsubscribeVehicleData request to HMI during the unregistration of App1
+-- 2) App_2 sends UnregisterAppInterface request to SDL
+-- SDL does:
+-- a) send UnsubscribeVehicleData request to HMI during the unregistration of App2
+---------------------------------------------------------------------------------------------------
```
Comment (dboltovskyi) [inline test_scripts/API/VehicleData/UnsubscribeVehicleData/011_UnsubscribeVD_for_2_apps_Expected_Disconnect_different_data.lua:21]: Since we have 2 VD parameters in this test I would add more details in description which app is subscribed to which parameter. And also which VD parameter should be in `UnsubscribeVehicleData` requests.
As we did in other `*_different_data` scripts.

File: test_scripts/API/VehicleData/SubscribeVehicleData/014_SubscribeVD_Resumption_of_subscription_for_2_apps_Ignition_Cycle_different_data.lua [hunk 12]
```diff
@@ -0,0 +1,64 @@
+---------------------------------------------------------------------------------------------------
+-- Description: Check that SDL resumes the subscription for different <vd_param> parameter for two Apps
```
Comment (dboltovskyi) [inline test_scripts/API/VehicleData/SubscribeVehicleData/014_SubscribeVD_Resumption_of_subscription_for_2_apps_Ignition_Cycle_different_data.lua:2]: Do we need to add info on issue as we did in other scripts?

**Review by dboltovskyi — CHANGES_REQUESTED** [review]

**Comment by VjKlepikov** [comment]
@dboltovskyi Thank you for the review, all comments have been processed

**Review by dboltovskyi — APPROVED** [review]
@VjKlepikov Thank you for updates!
